### PR TITLE
Don't use dynamic types for $apply implementation

### PR DESCRIPTION
**Pull request by kosinsky**
### Issues
*This pull request fixes issue #874.* 

### Description
Dynamic CLR types generated to represent shape in groupby/aggregate clauses are aggressively cached by Entity Framework 6 in many places and as result never garbage collected. As a result, each next call make system a bit slower

Re-implemented AggregationBinder using the same approach as SelectExpandBinder does (linked list of Containers). A bit of complexity was added that expression generated from each groupby could be use in further groupby, filter, orderby etc. expressions. It’s why we used dynamic types in first place, but recent refactoring in ODL v7/ WebApi v6 make linked list approach possible.

### Checklist (Uncheck if it is not completed)
- [ x ] Test cases added
- [ x ] Build and test with one-click build and test script passed

### CODE REVIEW

File: OData/src/System.Web.OData/OData/Formatter/Serialization/ODataResourceSerializer.cs [hunk 1]
```diff
@@ -235,31 +235,27 @@ private void WriteDeltaResource(object graph, ODataDeltaWriter writer, ODataSeri
         {
             var properties = new List<ODataProperty>();
             var dynamicObject = graph as DynamicTypeWrapper;
-            foreach (var prop in graph.GetType().GetProperties())
+            foreach (var prop in dynamicObject.Values)
             {
-                object value;
                 ODataProperty property = null;
```
Comment (xuzhg) [inline OData/src/System.Web.OData/OData/Formatter/Serialization/ODataResourceSerializer.cs]: move it into the inner loop

File: OData/src/System.Web.OData/OData/Formatter/Serialization/ODataResourceSerializer.cs [hunk 2]
```diff
@@ -235,31 +235,27 @@ private void WriteDeltaResource(object graph, ODataDeltaWriter writer, ODataSeri
...
+                        .FirstOrDefault(p => p.Name.Equals(prop.Key));
+                    if (edmProperty != null)
                     {
-                        IEdmProperty edmProperty = entityType.Properties()
-                            .FirstOrDefault(p => p.Name.Equals(prop.Name));
-                        if (edmProperty != null)
-                        {
-                            dynamicTypeProperties.Add(edmProperty, value);
-                        }
+                        dynamicTypeProperties.Add(edmProperty, prop.Value);
```
Comment (xuzhg) [inline OData/src/System.Web.OData/OData/Formatter/Serialization/ODataResourceSerializer.cs]: add assert for the dynamic type. [Resolved].

**Comment by xuzhg** [comment]
Merged with some changes at https://github.com/OData/WebApi/commit/7bcaedfebb949cdbefcb2cb9aa934a0d3a70d3f5.

@kosinsky Please verify the fix and let us know any problems.  

Thanks your contribution.

File: OData/src/System.Web.OData/OData/Query/Expressions/ExpressionBinderBase.cs [hunk 3]
```diff
@@ -511,6 +554,154 @@ internal Expression BindHas(Expression left, Expression flag)
...
+        /// <returns>Returns null if no aggregations were used so far</returns>
+        protected Expression GetFlattenedPropertyExpression(string propertyPath)
+        {
+            if (flattenedPropertyContainer == null)
+            {
+                return null;
+            }
+
+            Expression expression;
+            if (flattenedPropertyContainer.TryGetValue(propertyPath, out expression))
```
Comment (PietheinHeemskerk) [inline OData/src/System.Web.OData/OData/Query/Expressions/ExpressionBinderBase.cs]: The lines 696-700 should be replaced by the following lines to allow ODataConventionModelBuilderExtensions.EnableLowerCamelCase() to work. See issue #1136

```
#            var entry = FlattenedPropertyContainer.FirstOrDefault(c => string.Equals(c.Key, propertyPath, StringComparison.OrdinalIgnoreCase));
            if (entry.Key != null)
            {
                return entry.Value;
            }
```

**Pull request closed without merging**